lib/api/northbound-api.js

```diff
@@ -0,0 +1,211 @@
+// Copyright 2016, EMC, Inc.
+
+'use strict';
+
+var di = require('di'),
+    express = require('express');
+
+module.exports = NorthboundApiFactory;
+
+di.annotate(NorthboundApiFactory, new di.Provide('northbound-api'));
+di.annotate(NorthboundApiFactory, new di.Inject(
+    di.Injector,
+    'assert',
+    'Errors',
+    'Http.Services.RestApi',
+    'Services.Inventory',
+    'path',
+    'Logger',
+    'fs-operation',
+    'FileUploader',
+    'Services.Configuration',
+    'uuid',
+    '_'
+)
+);
+
+function NorthboundApiFactory(
+    injector,
+    assert,
+    Errors,
+    rest,
+    inventory,
+    path,
+    Logger,
+    fsOp,
+    Uploader,
+    configuration,
+    uuid,
+    _
+) {
+    var logger = Logger.initialize("northboundApi");
+    var router = express.Router();
+
+    var isoDirPath, microkernelDirPath;
+
+    function NorthboundApi(){
+    }
+
+    function uploadFile(req, res, filePath) {
+        var uploader = new Uploader(filePath);
+        return uploader.upload(req, res)
+            .then(function(){
+                if(fsOp.getFileSize(filePath) === 0){
+                    throw new Error('Error uploading file');
+                }
+            })
+            .catch(function(){
+                return fsOp.removeFile(filePath)
+                    .finally(function(){
+                        throw new Errors.BadRequestError('Error uploading file');
+                    });
+            });
+    }
+
+    function uploadIsoFile(req, res, iso) {
+        var isoPath = path.join(isoDirPath, iso);
+        return uploadFile(req, res, isoPath);
+    }
+
+    function uploadMicrokernelFile(req, res, iso) {
+        var microkernelPath = path.join(microkernelDirPath, iso);
+        return uploadFile(req, res, microkernelPath);
+    }
+
+    function createSymbolLinkInStore(localPath, isoName) {
+        if(fsOp.checkPathReadable(localPath)) {
+            var destPath = path.join(isoDirPath, isoName);
+
+            return fsOp.createSymbolLink(localPath, destPath)
+                .catch(function(){
+                    throw new Errors.InternalServerError(
+                        'Error creating symbol link to ISO file');
+                });
+        } else{
+            throw new Errors.BadRequestError("Path is not readable");
+        }
+    }
+
+    function getImageQuery(req){
+        var query = {};
+
+        query.id = req.query.id;
+        query.name = req.query.name;
+        query.version = req.query.version;
+        query.iso = req.query.iso;
+        query.status = req.query.status;
+
+        return _(query).omit(_.isUndefined).omit(_.isNull).value();
+    }
+
+    NorthboundApi.prototype.getRouter = function getRouter(){
+        isoDirPath = path.join(process.cwd(),
```

Comment:
The `process.cwd()` means the ISO folder must lie in the current service folder, I assume your intention is to let the folder can be anywhere in system, right? If so, you need some change here.